### PR TITLE
Add rnn() operator for MXNet backend with unrolling and masking feature

### PR DESCRIPTION
**Pull request by karan6181**
- Add run() operator in MXNet backend with unrolling and masking feature.
- Modified constant(), zeros_like(), and categorical_crossentropy() operator
- Enable test cases for run() operator with unrolling and masking feature.

@deep-learning-tools/aws-deep-learning-team @sandeep-krishnamurthy @roywei

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -2741,11 +2814,20 @@ def categorical_crossentropy(target, output, from_logits=False):
     # Returns
         Output tensor.
     """
-    assert not from_logits
     axis = ndim(output) - 1
     mx_output = output.symbol
+    # scale predictions so that the class probas of each sample sum to 1
+    if from_logits:
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Can you please enable categorical cross entropy tests.

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -3628,6 +3710,8 @@ def _get_type(self):
     def __getitem__(self, in_slice):
         begin = []
         end = []
+        if not isinstance(in_slice, (list, tuple)):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Why?

File: keras/backend/mxnet_backend.py [hunk 3]
```diff
@@ -2555,7 +2557,78 @@ def rnn(step_function, inputs, initial_states,
...
+    # with shape `(samples, ...)` (no time dimension)
+    inputs = list(mx.sym.split(inputs.symbol, axis=1,
+                               squeeze_axis=1, num_outputs=dshape[1]))
+
+    # Reverse the input sequence
+    if go_backwards:
+        inputs.reverse()
+
+    # Assume learning phase is a placeholder tensor.(F = test, T = train)
+    global uses_learning_phase
```
[the remark this answers is not in the document]
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: Some Keras layers (e.g. Dropout, BatchNormalization) behave differently at training time and testing time. uses_learning_phase is True if the layer has a different behavior in training mode and test mode, False otherwise.

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: `in_slice` should be a tuple or list of “slice()” constructor. But it fails(Single dimensional bias Tensors) if `in_slice` is not a tuple or list(Iterative object).

File: keras/backend/mxnet_backend.py [hunk 4]
```diff
@@ -274,7 +275,15 @@ def constant(value, dtype=None, shape=None, name=None):
         np_ndarray = np.ndarray(shape, dtype=dtype)
         np_ndarray.fill(value)
         mx_ndarray = mx.nd.array(np_ndarray)
-    return mx_ndarray
+
+    # If the Tensor shape is (1, ), then, it is considered to be scalar.
+    # Return the NDArray.
+    if shape == (1,):
+        return mx_ndarray
+    name = _prepare_name(name, 'constant')
```
[the remark this answers is not in the document]
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: Done. Thanks!

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -666,16 +676,7 @@ def zeros_like(x, dtype=None, name=None):
...
-        dtype = x.dtype
-    else:
-        dtype = _convert_string_dtype(dtype)
-    name = _prepare_name(name, 'zeroslikeinit')
-    mx_shape = tuple([0 if x is None else x for x in x.shape])
-    mx_value = mx.nd.zeros(mx_shape, dtype=dtype)
-    k_var = _keras_variable(name=name, dtype=dtype, shape=mx_shape)
-    k_var.bind(mx_value)
-    return k_var
+    return KerasSymbol(mx.sym.zeros_like(data=x.symbol))
```
[the remark this answers is not in the document]
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: Added the `name` parameter but the parameter`dtype` is optional now. MXNet symbolic `zeros_like()` internally generates an array of zeros with the same shape and type as the input array.

File: keras/backend/mxnet_backend.py [hunk 6]
```diff
@@ -2536,7 +2538,7 @@ def rnn(step_function, inputs, initial_states,
         constants: a list of constant values passed at each step.
         unroll: whether to unroll the RNN or to use a symbolic loop
         (`while_loop` or `scan` depending on backend).
-        input_length: not relevant in the TensorFlow implementation.
+        input_length: not relevant in the TensorFlow and MXNet implementation.
```
[the remark this answers is not in the document]
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: Done. Thanks!

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Please add this description as code comment.

File: keras/backend/mxnet_backend.py (hunk 1 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Comments?

Comment (karan6181) [reply keras/backend/mxnet_backend.py]: I enabled the categorical crossentropy testcases. Thanks!

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (karan6181) [reply keras/backend/mxnet_backend.py]: Done. Thanks!

**Comment by sandeep-krishnamurthy** [comment]
@karan6181 - Can you please handle `unroll=False` case appropriately with warnings/error message and providing a way for users simulate it with padding?

**Comment by sandeep-krishnamurthy** [comment]
@karan6181 - Can you please take this PR to completion?

**Comment by karan6181** [comment]
@sandeep-krishnamurthy Yes, working on a conv2d_transpose() operator use-cases due to modifications in one of the operators from my pull request. Once that's done, this PR is good to go to merge. Thank-You!.
